### ISSUE TITLE
`invlink(link)` should not be triggered in `avg_predictions()` or with `by`

### ISSUE THREAD

**Issue by vincentarelbundock** [vincentarelbundock/marginaleffects#1123]

**Comment by vincentarelbundock** [comment]
Counter-point: Russ Lenth does it in `emmeans`.

**Comment by ngreifer** [comment]
Reopening this because it has been bugging me a little. The fact that `emmeans` does it isn't a good reason to do it. The estimate doesn't correspond to any meaningful estimand. For a single prediction (i.e., no aggregating), the estimate is the same estimate as with `type = "response"` but we get a better confidence interval that respects the scale of the variable. When aggregating, the benefit is lost because the estimand changes (i.e., from the average of the predictions to the inverse link of the average of the linear predictor).

I think you already had this implemented here but changed it maybe?
https://github.com/vincentarelbundock/marginaleffects/blob/04848576d2c5d7281436aaebb740193f9696e85f/R/sanitize_type.R#L31

Let me know if there is any way I can convince you to do this.

**Comment by vincentarelbundock** [comment]
I'm inclined to agree with you that thinking of an average of predictions, rather than as the inverse link of the average-linear. But "doesn't correspond to any meaningful estimand" is way too strong. 

This is not just "what `emmeans` does". Lenth makes an impassioned pitch for this approach here: https://cran.r-project.org/web/packages/emmeans/vignettes/transformations.html

Again, I'm inclined to agree with you, but this is something reasonable people can disagree on. (Aside: I was at a conference last week and discussed this specific issue with D.Firth. I didn't come out with a definite answer, but what was obvious is that even highly respected statisticians don't have a clear cut answer to this problem.)

In terms of what it would take to do a breaking change, my desiderata would be:

1. An easy way to do both. 
  - A lot of people will be disturbed by intervals around `avg_predictions(logitmodel)` if they stretch outside [0,1]. 
  - There is value in having a very easy way to replicate `emmeans`, the main competitor.
  - With the old implementation you point to, `invlink(link)` gets dropped out of the type dictionary completely, which makes it impossible to use. So what's the alternative?
2. Good documentation. 
  - This is mentioned in `?predictions()`, as well as in the Predictions and the Standard Errors vignette. All that text should be corrected.
3. A note in NEWS with clear motivation for what is a breaking change.

**Comment by ngreifer** [comment]
Reading the `emmeans` vignette did help me see that there could be some cases where that approach has some validity, but in my opinion, those cases are still quite obscure. For example, if one fits a logistic regression model and then we compute the average of the logits (i.e., the linear predictor), we could interpret this as the log of the geometric mean of the predicted odds. But I don't know how taking the inverse logit of this quantity puts us back on a meaningful scale; it doesn't correspond to a probability even though it ranges from 0 to 1 and is a different quantity from the average predicted probability. One way it might be useful is if you want to interpret the linear predictor as a latent variable, so that the mean of the linear predictor is the mean of the latent variable underlying the choice. But again, the inverse logit of this latent variable doesn't correspond to a probability even though it will range from 0 to 1. I think the story could be different for transformations, which Lenth mentions, but I have not seen a solid argument for why to consider the inverse link of the average of the linear predictor. I would definitely be curious to hear what Dr. Firth had to say; it's awesome that you got to meet him!

Here is a solution that might be a little weird but it could be worth considering:

1. Compute the predictions on the response scale
2. Take the average (so far, just like `type = "response"`)
3. Apply the inverse link to this average to get a linear quantity (i.e., one with a range from `-Inf` to `Inf`)
4. Compute the standard errors (and CIs and tests) from this quantity
5. Backtransform the estimate and CI to the response scale

In base R, I'm imagining something like this:
```r
p <- predict(fit, newdata = newdata, type = "response")
mp <- mean(p) # ranges from 0 to 1
tmp <- qlogis(p) # ranges from -Inf to Inf
# Compute SEs on `tmp`
est <- plogis(tmp)
ci <- plogis(tmp - 1.96*se, tmp + 1.96*se)
```
This has the nice compromise of corresponding to the same estimand as when using `type = "response"` and still ensuring the CIs are within the domain of the outcome.

I'm not proposing `invlink(link)` is removed as I think it still has value for `predictions()` with `by` unspecified, in particular because it guarantees the intervals respect the bounds of the outcome; it's only for averaging predictions that I think this change should occur. If you change the meaning of `invlink(link)` to use the approach I describe above, you get the same result for individual predictions as for the current implementation; as Lenth mentions, it's just a matter of timing of transformation and averaging, and I'm just proposing switching the timing of the averaging from before the link to after the averaging. When no averaging takes place, the transformations occur in the same order they usually do.

This is analogous to what `avg_comparisons()` does with `contrast` and `transform`; the contrast occurs on the averaged response, SEs are computed on the contrast, and then the estimate is transformed by `transform`. That why for risk ratios, I set `contrast = "lnratioavg", transform = exp`; the log risk ratio is likely to be a linear quantity on which valid SEs can be computed, and the transformation puts the estimate and CI back on the desired scale. 

If this was implemented as new behavior, a user could request the same output `emmeans` produces by using `avg_predictions(., type = "link", transform = plogis)`, which is I think what `invlink(link)` is a shortcut for anyway. Using `predictions()` without `by` would not change. The only issue is that the user needs to know what the inverse link function is to supply to `transform`. One option would be to have a keyword, like `".link"`, that could be supplied to `transform` to automatically trigger the inverse link. I think because this is a more esoteric quantity (I argue), it's okay if the user has to do a little more work to request it. I think a danger is the user thinking the `invlink(link)` default is actually producing what `"response"` produces; I know I have been confused by this since it only occurs with `avg_predictions()` and not `avg_comparisons()`.

**Comment by vincentarelbundock** [comment]
> it might be useful is if you want to interpret the linear predictor as a latent variable

Right, in my head I kept hearing the word "generative", but that's probably a better characterization. I do think it is a good strategy to pitch `marginaleffects` more directly at the kinds of quantities in your "causal estimands" paper, which often involve $E[Y]$. So it's natural to take the mean of the predictions directly on the response scale.

Your proposed solution is interesting, but I worry that it's too complicated (and also something I haven't seen before). I think I'd prefer something that is easier to explain and document:

1. If you call `predictions(..., by = FALSE)` with a `glm` or `gam()` model, we predict on the link scale and back transform.
2. In all other cases, we predict on the response scale, or on the scale explicitly selected by the user with the `type` argument.

That's a clear cut rule. Easy to explain for us. Easy to understand for users.

The tricky part would be to allow:

```r
avg_predictions(model, type = "invlink(link)")
```

Currently, the commented out code that you linked to makes this impossible because `invlink(link)` is dropped out of the type dictionary completely. So that `if` condition needs to be amended to avoid dropping those lines from the type dict when the user has explicitly set the `type` argument when calling the function. 

The other piece of work we'd need to get done is documentation. There are the three places noted above in thread, and also the NEWS.

**Comment by ngreifer** [comment]
I think what you proposed is perfect and honestly what I would have expected the default behavior was anyway. I think you can move the block of code that removes `invlink(link)` from the type dictionary to only activate when `type` is `NULL` (i.e., after line 45).

My idea is definitely an esoteric one, but I do think it could have nice properties. I agree with your reasoning to avoid it, though. Using quantile intervals from simulation-based inference is equivalent to doing this but doesn't require knowledge of which transformation is required to get a valid symmetric CI before transforming (and doesn't require the transformation).

**Comment by vincentarelbundock** [comment]
Fixed on github
